server/app/services/scheduler/filter/memory.rb

```diff
@@ -30,25 +27,57 @@ def for_service(service, instance_number, nodes)
         candidates
       end
 
+      def resolve_memory_from_stats(service, instance_number)
+        cache_key = "scheduler:mem:filter:#{service.id}-#{instance_number}"
+        memory = self.class.cache[cache_key]
+        unless memory
+          container = service.containers.to_a.find { |c| c.instance_number == instance_number }
+          if container
+            stats = container.container_stats.latest
+            if stats
+              memory = stats.memory['usage'] * 1.25
+              self.class.cache[cache_key] = memory
+            end
+          end
+        end
+        unless memory
+          memory = 256.megabytes # just a default if we cannot resolve anything
```

Comment:
Interesting. Does this mean that we will now refuse to schedule any new service instances onto a node that has less than 256mb of free memory?

This PR isn't just a perf optimization anymore... it's also changing the behavior? Should this change be in a separate follow-up PR?

Reply:
I can split this into separate pr if needed. I just feel that scheduling without any mem limit is a bug and can hit users too many ways.

Reply:
> Does this mean that we will now refuse to schedule any new service instances onto a node that has less than 256mb of free memory?

Yes if service does not define a memory limit or it does not have any stats.